File: src/main/java/nz/zephire/exceptions/StateConfig.java
package nz.zephire.exceptions;

import com.google.common.base.Charsets;
import com.google.gson.Gson;
import nz.zephire.exceptions.exceptions.Exception;
import nz.zephire.exceptions.exceptions.PreApprovedException;
import org.joda.time.LocalTime;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

/**
 * Holds persisted state information, and saves/loads to ensure state is transferred across restarts.
 */
public class StateConfig {

    public static final transient Path SAVE_LOCATION = Paths.get(System.getProperty("user.home"),".config", "exceptions.conf");

    private static transient Gson gson = Utils.getGson();

    private String username;

    private LocalTime startTime;

    private LocalTime endTime;

    private boolean isFinished = true;

    private Set<PreApprovedException> preApproved = new HashSet<>();

    private Set<Exception> exceptions = new HashSet<>();

    //TODO: Exceptions

    public static StateConfig loadConfig() {
        try {
            return gson.fromJson(Files.newBufferedReader(SAVE_LOCATION, Charsets.UTF_8), StateConfig.class);
        } catch (IOException e) {
            Utils.debugLog("Unable to load config!");
            Utils.debugLog(e);
            throw new RuntimeException(e);
        }
    }

    public void saveConfig() {
        try {
            gson.toJson(this, Files.newBufferedWriter(SAVE_LOCATION, Charsets.UTF_8, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING));
        } catch (IOException e) {
            Utils.debugLog("Unable to save config!");
            Utils.debugLog(e);
            Utils.debugLog("");
            Utils.debugLog("State at save:");
            Utils.debugLog(this.toString());
            throw new RuntimeException(e);
        }
    }

    public String getUsername() {
        return username;
    }

    public LocalTime getStartTime() {
        return startTime;
    }

    public LocalTime getEndTime() {
        return endTime;
    }

    public boolean isFinished() {
        return isFinished;
    }

    public Set<PreApprovedException> getPreApproved() {
        return Collections.unmodifiableSet(preApproved);
    }

    public Set<Exception> getExceptions() {
        return Collections.unmodifiableSet(exceptions);
    }

    public void setUsername(String username) {
        this.username = username;
        saveConfig();
    }

    public void setStartTime(LocalTime startTime) {
        this.startTime = startTime;
        saveConfig();
    }

    public void setEndTime(LocalTime endTime) {
        this.endTime = endTime;
        saveConfig();
    }

    public void setFinished(boolean finished) {
        this.isFinished = finished;
        saveConfig();
    }

    public void addPreApproved(PreApprovedException preApproved) {
        this.preApproved.add(preApproved);
        saveConfig();
    }

    public void addException(Exception exception) {
        this.exceptions.add(exception);
        saveConfig();
    }

    @Override
    public String toString() {
        return "StateConfig{" +
                "username='" + username + '\'' +
                ", startTime=" + startTime +
                ", endTime=" + endTime +
                '}';
    }
}

File: src/main/java/nz/zephire/exceptions/StateConfig.kt
package nz.zephire.exceptions

import com.google.common.collect.Lists
import com.google.gson.annotations.SerializedName
import nz.zephire.exceptions.exceptions.Exception
import nz.zephire.exceptions.exceptions.RosteredException
import org.joda.time.LocalDate
import org.joda.time.LocalTime
import java.io.File
import java.io.IOException
import java.nio.file.Paths
import java.util.*

val SAVE_LOCATION: File = Paths.get(System.getProperty("user.home"), ".config", "exceptions.conf").toFile()

private val GSON = Utils.getGson()

fun loadConfig(): StateConfig {
    try {
        if(!SAVE_LOCATION.exists()) {
            val conf = StateConfig()
            conf.saveConfig()
            return conf
        }

        return GSON.fromJson(SAVE_LOCATION.bufferedReader(), StateConfig::class.java)
    } catch (e: IOException) {
        Utils.debugLog("Unable to load config!")
        Utils.debugLog(e)
        throw RuntimeException(e)
    }
}

/**
 * Holds persisted state information, and saves/loads to ensure state is transferred across restarts.
 */
class StateConfig {

    var username: String = ""
        set(username) {
            field = username
            saveConfig()
        }

    var startTime: LocalTime = LocalTime.MIDNIGHT
        set(value) {
            field = value
            saveConfig()
        }

    var endTime: LocalTime = LocalTime.MIDNIGHT
        set(value) {
            field = value
            saveConfig()
        }

    var previousDay: LocalDate = LocalDate.now()
        set(value) {
            field = value
            saveConfig()
        }

    var isFinished = true
        set(value) {
            field = value
            saveConfig()
        }

    var currentExceptionStartTime: LocalTime? = null
        set(value) {
            field = value
            saveConfig()
        }

    @SerializedName("pre_approved")
    private var _rostered: MutableList<RosteredException> = Lists.newArrayList()

    @Transient
    val rostered: List<RosteredException> = Collections.unmodifiableList(_rostered)

    @SerializedName("exceptions")
    private var _exceptions: MutableList<Exception> = Lists.newArrayList()

    @Transient
    val exceptions: List<Exception> = Collections.unmodifiableList(_exceptions)

    @SerializedName("current_rostered")
    private var _currentRostered: Int = -1 // Reference to list of Pre-Approved, or -1 for none
        set(value) {
            field = value
            saveConfig()
        }

    @Transient
    var currentRostered: RosteredException? = rostered.getOrNull(_currentRostered)

    fun addRosteredException(rostered: RosteredException) {
        this._rostered.add(rostered)
        saveConfig()
    }

    fun addException(exception: Exception) {
        this._exceptions.add(exception)
        saveConfig()
    }

    fun saveConfig() {
        val out = SAVE_LOCATION.bufferedWriter()

        try {
            GSON.toJson(this, out)
        } catch (e: IOException) {
            Utils.debugLog("Unable to save config!")
            Utils.debugLog(e)
            Utils.debugLog("")
            Utils.debugLog("State at save:")
            Utils.debugLog(this.toString())
            throw RuntimeException(e)
        } finally {
            out.close()
        }

    }

    override fun toString(): String {
        return "StateConfig(" +
                "username='$username', " +
                "startTime=$startTime, " +
                "endTime=$endTime, " +
                "previousDay=$previousDay, " +
                "isFinished=$isFinished, " +
                "currentExceptionStartTime=$currentExceptionStartTime, " +
                "rostered=$rostered, " +
                "exceptions=$exceptions" +
                ")"
    }
}
